src/libnnstreamer-edge/nnstreamer-edge-aitt.c

```diff
@@ -23,11 +28,218 @@ typedef void *nns_edge_aitt_sub_h;
+  }
+  aitt_destroy (ah->aitt_h);
+  ah->aitt_h = NULL;
+  SAFE_FREE (eh->broker_h);
+  eh->broker_h = NULL;
+
+  return NNS_EDGE_ERROR_NONE;
+}
+
+/**
+ * @brief Publish raw data.
+ * @note This is internal function forAITT. You should call this with edge-handle lock.
+ */
+int
+nns_edge_aitt_publish (nns_edge_h edge_h, const void *data, const int length)
+{
+  nns_edge_handle_s *eh;
+  nns_edge_aitt_handle_s *ah;
+
+  eh = (nns_edge_handle_s *) edge_h;
+
+  if (!NNS_EDGE_MAGIC_IS_VALID (eh) || !eh->broker_h) {
+    nns_edge_loge ("Invalid param, given edge handle is invalid.");
+    return NNS_EDGE_ERROR_INVALID_PARAMETER;
+  }
+
+  if (!data || length <= 0) {
+    nns_edge_loge ("Invalid param, given data is invalid.");
+    return NNS_EDGE_ERROR_INVALID_PARAMETER;
+  }
+
+  ah = (nns_edge_aitt_handle_s *) eh->broker_h;
+
+  if (AITT_ERROR_NONE != aitt_publish (ah->aitt_h, eh->topic, data, length)) {
+    nns_edge_loge ("Failed to publish the message. topic: %s", eh->topic);
+    return NNS_EDGE_ERROR_IO;
+  }
+
+  return NNS_EDGE_ERROR_NONE;
+}
+
+/**
+ * @brief Internal function to invoke event callback.
+ * @note This function should be called with handle lock.
+ */
+static int
+_nns_edge_invoke_event_cb (nns_edge_handle_s * eh, nns_edge_event_e event,
+    void *data, size_t data_len, nns_edge_data_destroy_cb destroy_cb)
+{
+  nns_edge_event_h event_h;
+  int ret;
+
+  /* If event callback is null, return ok. */
+  if (!eh->event_cb) {
+    nns_edge_logw ("AITT: The event callback is null, do nothing!");
+    return NNS_EDGE_ERROR_NONE;
+  }
+
+  ret = nns_edge_event_create (event, &event_h);
+  if (ret != NNS_EDGE_ERROR_NONE) {
+    nns_edge_loge ("Failed to create new edge event.");
+    return ret;
+  }
+
+  if (data) {
+    ret = nns_edge_event_set_data (event_h, data, data_len, destroy_cb);
+    if (ret != NNS_EDGE_ERROR_NONE) {
+      nns_edge_loge ("Failed to handle edge event due to invalid event data.");
+      goto error;
+    }
+  }
+
+  ret = eh->event_cb (event_h, eh->user_data);
+  if (ret != NNS_EDGE_ERROR_NONE) {
+    nns_edge_loge ("The event callback returns error.");
+  }
+
+error:
+  nns_edge_event_destroy (event_h);
+  return ret;
+}
+
+/**
+ * @brief Callback function to be called when a message is arrived.
+ */
+static void
+aitt_cb_message_arrived (aitt_msg_h msg_handle, const void *msg,
+    size_t msg_len, void *user_data)
+{
+  nns_edge_handle_s *eh;
+  nns_edge_data_h data_h;
+
+  eh = (nns_edge_handle_s *) user_data;
+
+  if (!NNS_EDGE_MAGIC_IS_VALID (eh)) {
+    nns_edge_loge ("Invalid param, given edge handle is invalid.");
+    return;
+  }
+
+  if (nns_edge_data_create (&data_h) != NNS_EDGE_ERROR_NONE) {
+    nns_edge_loge ("Failed to create data handle in msg thread.");
+    return;
+  }
+
+  /** @todo support multi memory chunk. Deserialize the received data. */
+  nns_edge_data_add (data_h, (void *) msg, msg_len, NULL);
+
+  _nns_edge_invoke_event_cb (eh, NNS_EDGE_EVENT_NEW_DATA_RECEIVED, data_h,
```

Comment:
Q: Does it need destroying data_h after calling event callback? (Now data handle should be copied if it is necessary.)